sdk/core/System.ClientModel/src/Convenience/ClientCache.cs

```diff
@@ -34,14 +34,14 @@ public ClientCache(int maxSize = 100)
     /// </summary>
     /// <typeparam name="T">The type of the client.</typeparam>
     /// <param name="createClient">A factory function to create the client if not cached.</param>
-    /// <param name="id">An identifier for the client instance.</param>
+    /// <param name="key">An equality-comparable key representing the client configuration.</param>
     /// <returns>The cached or newly created client instance.</returns>
-    public T GetClient<T>(Func<T> createClient, string? id) where T : class
+    public T GetClient<T>(Func<T> createClient, IEquatable<object>? key) where T : class
     {
-        (Type, string) key = (typeof(T), id ?? string.Empty);
+        var cacheKey = (typeof(T), key);
```

Comment:
let's not wrap type and key into this tuple and instead just use the key to lookup the dictionary. Then simply fail is the value int he dictionary cannot be cast to T